File: docs/case-studies/09-distributed-cache/go/consistent_hash_ring.go
// Scenario: consistent hashing ring lookup for distributed cache
// Demonstrates: key-to-node mapping and ring wrap-around
// Trade-off: simple hashing versus virtual node balancing

package cache

type ConsistentHashRing struct {
	points []int
	nodes  map[int]string
}

func (r *ConsistentHashRing) LocateNode(key string) string {
	hash := hashFn(key)
	for _, point := range r.points {
		if point >= hash {
			return r.nodes[point]
		}
	}
	return r.nodes[r.points[0]]
}

func hashFn(key string) int {
	h := 0
	for _, c := range key {
		h = 31*h + int(c)
	}
	return h & 0x7fffffff
}

File: case-studies/09-distributed-cache/java/ConsistentHashRing.java
// Scenario: consistent hashing ring for distributed cache
// Demonstrates: node lookup for a key and wrap-around behavior
// Trade-off: even distribution versus virtual node complexity

package case_studies.cache;

import java.util.NavigableMap;
import java.util.TreeMap;

public class ConsistentHashRing {
    private final NavigableMap<Integer, String> ring = new TreeMap<>();

    public void addNode(String nodeId, int hash) {
        ring.put(hash, nodeId);
    }

    public String locateNode(String key) {
        int hash = key.hashCode();
        if (ring.containsKey(hash)) {
            return ring.get(hash);
        }
        var tail = ring.tailMap(hash, false);
        if (!tail.isEmpty()) {
            return tail.firstEntry().getValue();
        }
        return ring.firstEntry().getValue();
    }
}
